In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from plotly.offline import init_notebook_mode, iplot, plot
import plotly as py
init_notebook_mode(connected=True)
import plotly.graph_objs as go
import plotly.express as px


import matplotlib.pyplot as plt
import os

# Требования: 
- во всех графиках должны быть подписи чере title, legend, etc... 
- Убедитесь, что после сохранения ноутбука графики всё ещё видно, если открыть ноутбук заново, если не видно, то напишите мне @kirizhik 
- Можно баловаться с цветами, чтобы было **K P A C U B O**, но **не вырвеглазно**, пожалуйста

In [53]:
df = pd.read_csv('/content/sample_data/data.csv', index_col = 0)

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18207 entries, 0 to 18206
Data columns (total 88 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        18207 non-null  int64  
 1   Name                      18207 non-null  object 
 2   Age                       18207 non-null  int64  
 3   Photo                     18207 non-null  object 
 4   Nationality               18207 non-null  object 
 5   Flag                      18207 non-null  object 
 6   Overall                   18207 non-null  int64  
 7   Potential                 18207 non-null  int64  
 8   Club                      17966 non-null  object 
 9   Club Logo                 18207 non-null  object 
 10  Value                     18207 non-null  object 
 11  Wage                      18207 non-null  object 
 12  Special                   18207 non-null  int64  
 13  Preferred Foot            18159 non-null  object 
 14  Intern

In [52]:
df

ID                Name  Age  \
0      158023            L. Messi   31   
1       20801   Cristiano Ronaldo   33   
2      190871           Neymar Jr   26   
3      193080              De Gea   27   
4      192985        K. De Bruyne   27   
...       ...                 ...  ...   
18202  238813        J. Lundstram   19   
18203  243165  N. Christoffersson   19   
18204  241638           B. Worman   16   
18205  246268      D. Walker-Rice   17   
18206  246269           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png        9         94   
1      https://cdn.sofifa.org/flags/38.png        9         94   
2      https://cdn.sofifa.org/flags/54.png        9         93   
3      https://cdn.sofifa.org/flags/45.png        9         93   
4       https://cdn.sofifa.org/flags/7.png        9         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png        4         65   
18203  https://cdn.sofifa.org/flags/46.png        4         63   
18204  https://cdn.sofifa.org/flags/14.png        4         67   
18205  https://cdn.sofifa.org/flags/14.png        4         66   
18206  https://cdn.sofifa.org/flags/14.png        4         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

       ... Composure Marking  StandingTackle SlidingTackle  GKDiving  \
0      ...      96.0    33.0            28.0          26.0       6.0   
1      ...      95.0    28.0            31.0          23.0       7.0   
2      ...      94.0    27.0            24.0          33.0       9.0   
3      ...      68.0    15.0            21.0          13.0      90.0   
4      ...      88.0    68.0            58.0          51.0      15.0   
...    ...       ...     ...             ...           ...       ...   
18202  ...      45.0    40.0            48.0          47.0      10.0   
18203  ...      42.0    22.0            15.0          19.0      10.0   
18204  ...      41.0    32.0            13.0          11.0       6.0   
18205  ...      46.0    20.0            25.0          27.0      14.0   
18206  ...      43.0    40.0            43.0          50.0      10.0   

       GKHandling  GKKicking GKPositioning GKReflexes Release Clause  
0          

In [44]:
df.columns

Index(['ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag', 'Overall',
       'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingTackle', 'GKDiv

In [45]:
df[['Wage']].head(10)

Wage
0  €565K
1  €405K
2  €290K
3  €260K
4  €355K
5  €340K
6  €420K
7  €455K
8  €380K
9   €94K

# Заполните пропуски в утверждениях 

- В датасете 88 переменных 

- Из них числовых 43

# Изучите распределение возростов с помощью Age [px.histogram](https://plotly.com/python/histograms/), изучите [более продвинутые методы](https://plotly.com/python/distplot/) визуализации распределений 

In [7]:
fig = px.histogram(df, x="Age", title="Age distribution")
fig.show(renderer="colab")

# Визуализируйте соотношенеие между левшами и правшами (? сами найдите необходимую колонку в данных, считаем, что левша-правша ~ предпочтительная нога для ударов )

In [8]:
fig = px.histogram(df, x="Preferred Foot", title="Ratio between preferred foot")
#fig.update_traces(marker_color='red')
fig.show(renderer="colab")


Правшей намного больше

# Визуализируйте Международный Рейтинг футболистов (International Reputation), ассоциирован ли он с национальностью? футбольным клубом? 

![Игроки](Players.png "Title")


In [9]:
fig = px.histogram(df, x="International Reputation", color='Nationality', title="International Reputation distribution")
#fig.update_traces(marker_color='red')
fig.show(renderer="colab")


Большая часть футболистов с репутацией равной 1

Для выяснения связи между национальностью и рейтингом имеет смысл отсеять игроков с рейтингом равным 1, так как их слишком много, и посмотреть на зависимость рейтинга от национальности для оставшихся игроков.

In [10]:
df_new = df.groupby('Nationality')['International Reputation'].value_counts().reset_index(name='counts')
df_new = df_new[df_new['International Reputation'] >=2]
df_new = df_new.groupby('Nationality')['counts'].sum().reset_index(name='all_counts')
df_new.sort_values(by='all_counts',ascending=False,inplace=True)

In [11]:
df_new

Nationality  all_counts
35         France         149
82          Spain         139
38        Germany         130
47          Italy         124
3       Argentina          94
..            ...         ...
51          Kenya           1
33  FYR Macedonia           1
49        Jamaica           1
45           Iran           1
67           Oman           1

[95 rows x 2 columns]

In [12]:
fig = px.bar(df_new, x="Nationality", y="all_counts", title="International Reputation distribution more then 2")
fig.show(renderer="colab")

Как видно, большая часть игроков с большим рейтингом распределена по 8 странам

Посмотрим, как обстоят дела с футбольными клубами(возьмем клубы с количеством игроков больше 20).

In [13]:
ir_best_avg=df.groupby('Club').apply(lambda x:np.mean(x['International Reputation'])).reset_index(name='Average Reputation')
ir_best_cnt=df.groupby('Club').apply(lambda x:x['Name'].count()).reset_index(name='Player Counts')
ir_best_avg_cnt=pd.merge(ir_best_avg,ir_best_cnt,how='inner',left_on='Club',right_on='Club')
ir20_best_avg_cnt=ir_best_avg_cnt[ir_best_avg_cnt['Player Counts']>=20]
ir20_best_avg_cnt.sort_values(by=['Average Reputation','Player Counts'],ascending=[False,False])
fig = px.scatter(ir20_best_avg_cnt,x='Average Reputation',y='Player Counts',color='Player Counts',size='Average Reputation',hover_data=['Club'],title='Clubwise player counts and Average Reputation')
fig.show(renderer="colab")

Да, репутация связана с клубом, можно выделить несколько клубов, где средний рейтинг игроков ощутимо больше чем у всех остальных

# Есть ли зависимость между потенциалом (Potential) и Международным Рейтингом футболистов (International Reputation)? Визуализируте её. Влияет ли на потенциал игровая нога? Почему так может случиться? (картинка в тему)

![Коазуация или корреляция?](https://github.com/kryzhikov/Sample-ML-Repo/blob/main/02.%20Data%20Wrangling/fifaeda/causuation.png?raw=1 "???")

In [14]:
fig = px.box(df, x="International Reputation", y="Potential", title="Relationship between potential and International Reputation")
fig.show(renderer="colab")

Да, зависимость есть. Правда, наверное, не совсем корректно учитывать игроков с потенциалом выше трех, так как их слишком мало.

In [15]:
fig = px.box(df, x="Preferred Foot", y="Potential", title="Relationship between potential and  Preferred Foot")
fig.show(renderer="colab")

Игровая нога на потенциал никак влияет(впрочем, ожидаемый результат)

# Теперь пройдёмся по показателям игроков... 
### Есть ли зависимость Выносливости (Stamina) и Силы (Strength) игроков? какая? точно зависимость? 


In [16]:
fig = px.bar(df, x="Stamina", y="Strength", title="Relationship between Stamina and Strength")
fig.show(renderer="colab")

Построим график зависимости медианного значения Strength для заданного значения Stamina:

In [17]:
df_new = df.groupby('Stamina').apply(lambda x: x['Strength'].median()).reset_index(name='Median Strength')

In [18]:
fig=px.scatter(df_new, x='Stamina',y='Median Strength',title='Dependence of Stamina on median strength')
fig.show(renderer='colab')

Да, возможно есть линейная зависимость, но крайне слабая

# Зависит ли Стоимость игрока стоимость (Value) игрока от его "финтов" (Skill Moves)? 

In [7]:
df_new = df[['Value', 'Skill Moves']]

In [9]:
df_new['MultV']=np.where(df_new.Value.str[-1]=='K',1000,np.where(df_new.Value.str[-1]=='M',1000000,1))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [10]:
df_new['Value']=df_new.Value.str.strip('€')
df_new['Value']=df_new.Value.str.strip('K')
df_new['Value']=df_new.Value.str.strip('M')
df_new['Value']=pd.to_numeric(df_new['Value'])
df_new['Value in €']=df_new['Value']*df_new['MultV']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/st

In [11]:
df_new

Value  Skill Moves    MultV   Value in €
0      110.5          4.0  1000000  110500000.0
1       77.0          5.0  1000000   77000000.0
2      118.5          5.0  1000000  118500000.0
3       72.0          1.0  1000000   72000000.0
4      102.0          4.0  1000000  102000000.0
...      ...          ...      ...          ...
18202   60.0          2.0     1000      60000.0
18203   60.0          2.0     1000      60000.0
18204   60.0          2.0     1000      60000.0
18205   60.0          2.0     1000      60000.0
18206   60.0          2.0     1000      60000.0

[18207 rows x 4 columns]

In [12]:
df_new = df_new.groupby('Skill Moves').apply(lambda x: x['Value in €'].median()).reset_index(name='Median Value in €')

In [14]:
fig=px.scatter(df_new, x='Skill Moves',y='Median Value in €',title='Dependence of Value on Skill Moves')
fig.show(renderer='colab')


Дв, наблюдается явная зависимость.

# Покажите распределеняи основных характеристик игроков: Рост, Вес, Возраст по позициям (Position), сделайте выводы...


Для начала было бы интересно посмотреть на распределение игроков по позициям(возьмем 20 самых популярных):

In [67]:
df_new = df.groupby('Position').apply(lambda x:x['Name'].count()).reset_index(name='Counts')
df_new.sort_values(by='Counts',ascending=False,inplace=True)
df_new_top_20 = df_new[:20]
fig=px.bar(df_new_top_20, x='Position', y='Counts', color='Counts', title='Positionwise Player counts')
fig.show(renderer='colab')

Придется переводить футы и дюймы в сантиметры(

In [68]:
df_new = df[df['Height'].notna()]

In [69]:
def height(x):
  #list_with = list(x.split('\''))
  a = float(list(x.split('\''))[0])
  b = float(list(x.split('\''))[1])
  return a * 30.48 / 100 + b * 2.54 / 100

In [70]:
df_new['Height in metres'] = df_new['Height'].apply(height)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [71]:
df_new_pos_height = df_new.groupby('Position').apply(lambda x: x['Height in metres'].median()).reset_index(name='Median Height')

In [72]:
df_new_pos_height.sort_values('Median Height', inplace=True)

In [73]:
fig=px.scatter(df_new_pos_height, x='Position',y='Median Height',title='Dependence of Position on median Height')
fig.show(renderer='colab')

Можно сделать вывод о том, что для некоторых позиций рост важен(например для защитников нужен высокий рост)

Посмотрим на зависимость веса от позиции:

In [74]:
df_new = df[df['Weight'].notna()]

In [75]:
def weight(x):
  #list_with = list(x.split('l'))
  a = int(list(x.split('l'))[0])
  #b = float(list(x.split('\''))[1])
  return a 

In [76]:
df_new['Weight in int'] = df_new['Weight'].apply(weight)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [77]:
df_new_pos_weight = df_new.groupby('Position').apply(lambda x: x['Weight in int'].median()).reset_index(name='Median Weight')

In [78]:
df_new_pos_weight.sort_values('Median Weight', inplace=True)

In [79]:
fig=px.scatter(df_new_pos_weight, x='Position',y='Median Weight',title='Dependence of Position on median Weight')
fig.show(renderer='colab')

Наблюдается явная зависимость веса от позиции

Проанализируем распределение по возрасту:

In [80]:
df_new = df[df['Age'].notna()]

In [81]:
df_new_pos_age = df_new.groupby('Position').apply(lambda x: x['Age'].median()).reset_index(name='Median Age')

In [82]:
df_new_pos_age.sort_values('Median Age', inplace=True)

In [83]:
fig=px.scatter(df_new_pos_age, x='Position',y='Median Age',title='Dependence of Position on median Age')
fig.show(renderer='colab')

Да. для возраста тоже наблюдается явная зависимость

# Есть ли зависимость скорости (SprintSpeed) от веса и роста? А от национальности?  

Для начала глянем частоту с которой встречается скорость:

In [38]:
fig=px.histogram(df, x='SprintSpeed',title='Dependence of Position on median Age')
fig.show(renderer='colab')

In [41]:
df_new = df[df['Weight'].notna()]

In [42]:
def weight(x):
  #list_with = list(x.split('l'))
  a = int(list(x.split('l'))[0])
  #b = float(list(x.split('\''))[1])
  return a 

In [43]:
df_new['Weight in int'] = df_new['Weight'].apply(weight)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [44]:
df_new_speed_weight = df_new.groupby('SprintSpeed').apply(lambda x: x['Weight in int'].median()).reset_index(name='Median Weight')

In [45]:
df_new_speed_weight.sort_values('Median Weight', inplace=True)

In [46]:
fig=px.scatter(df_new_speed_weight, x='SprintSpeed',y='Median Weight',title='Dependence of SprintSpeed on median Weight')
fig.show(renderer='colab')

Чем больше вес, тем меньше скорость - предсказуемый результат(крайние точки видимо выбросы)

Проделаем то же самое для роста:

In [48]:
df_new = df[df['Height'].notna()]


In [49]:
def height(x):
  #list_with = list(x.split('\''))
  a = float(list(x.split('\''))[0])
  b = float(list(x.split('\''))[1])
  return a * 30.48 / 100 + b * 2.54 / 100

In [50]:
df_new['Height in metres'] = df_new['Height'].apply(height)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [53]:
df_new_speed_height = df_new.groupby('SprintSpeed').apply(lambda x: x['Height in metres'].median()).reset_index(name='Median Height')

In [54]:
df_new_speed_height.sort_values('Median Height', inplace=True)

In [55]:
fig=px.scatter(df_new_speed_height, x='SprintSpeed',y='Median Height',title='Dependence of SprintSpeed on median Height')
fig.show(renderer='colab')

Также есть зависимость - чем меньше рост, тем больше SprintSpeed

Посмотрим на национальности:

In [57]:
df_new = df.groupby('Nationality').apply(lambda x: x['SprintSpeed'].median()).reset_index(name='Median SprintSpeed')

In [71]:
df_new.sort_values('Median SprintSpeed', ascending=False, inplace=True)

In [72]:
df_new_top_20 = df_new[:30]

In [73]:
fig=px.scatter(df_new_top_20, x='Nationality', y='Median SprintSpeed', color='Median SprintSpeed', title='Dependence of SprintSpeed on Nationality')
fig.show(renderer='colab')

Как видно, весь топ занят африканцами

# Под какими номерами чаще всего играют нападающие? Визуализируйте частотность  

In [ ]:
df_new =df[df['Position'] == 'ST']

In [ ]:
fig = px.histogram(df, x="Jersey Number", title="ST numbers distribution")
fig.show(renderer="colab")

Чаще всего играют под номерами от 1 до 11

# Есть ли за зависимость **силы удара** (ShotPower) от **ведущей ноги**? от **агрессивности** (Aggression)? 


In [ ]:
df_new = df.drop(['Loaned From'], axis=1)
df_new = df_new.dropna()
fig = px.histogram(df_new, x="ShotPower", color='Preferred Foot', title="Dependency ShotPower on Preferred Foot")
fig.show(renderer="colab")

Нет, зависимости от ведущей ноги нет, распределения практически одинаковы.

Посмотрим на зависимость от агрессивности:

In [ ]:
df_new = df.groupby('ShotPower').apply(lambda x: x['Aggression'].median()).reset_index(name='Median Aggression')

In [ ]:
fig=px.scatter(df_new, x='ShotPower',y='Median Aggression',title='Dependency ShotPower on median Aggression')
fig.show(renderer='colab')

Зависимость получилась странная, но вроде бы да, зависит.

# Найдите суммарную стоимость игроков в каждом клубе. Визуализируйте соотношение сумарной стоимости топ 10 клубов. 

In [3]:
df_new = df[['Name', 'Club', 'Value']]

In [4]:
df_new['MultV']=np.where(df_new.Value.str[-1]=='K',1000,np.where(df_new.Value.str[-1]=='M',1000000,1))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [5]:
df_new['Value']=df_new.Value.str.strip('€')
df_new['Value']=df_new.Value.str.strip('K')
df_new['Value']=df_new.Value.str.strip('M')
df_new['Value']=pd.to_numeric(df_new['Value'])
df_new['Value in €']=df_new['Value']*df_new['MultV']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/st

In [13]:
df_new = df_new.groupby('Club').apply(lambda x: np.sum(x['Value in €'])).reset_index(name='Summary value')

In [7]:
df_new_name = df[['Club', 'Name']]

In [8]:
df_new_name = df_new_name.groupby('Club').apply(lambda x: x['Name'].count()).reset_index(name='Counts')

In [14]:
df_all = pd.merge(df_new, df_new_name, how='inner',left_on='Club',right_on='Club')

In [15]:
df_all

Club  Summary value  Counts
0      SSV Jahn Regensburg     23095000.0      29
1    1. FC Heidenheim 1846     27490000.0      28
2     1. FC Kaiserslautern     16195000.0      26
3               1. FC Köln    135910000.0      28
4          1. FC Magdeburg     20525000.0      26
..                     ...            ...     ...
646     Zagłębie Sosnowiec      6815000.0      25
647        Çaykur Rizespor     51840000.0      30
648              Örebro SK     10685000.0      27
649          Östersunds FK     13180000.0      22
650          Śląsk Wrocław      9595000.0      25

[651 rows x 3 columns]

Сверху суммарная стоимость игроков в каждом клубе.

In [17]:
df_all.sort_values(['Summary value'], ascending=False, inplace=True)

In [19]:
df_all = df_all[:10]

In [21]:
fig = px.bar(df_all, x="Club", y='Summary value', color='Counts', title="Top 10 clubs by value")
fig.show(renderer="colab")

Суммарная стоимость топ 10 клубов варьируется в пределах 25 процентов

# Есть ли зависимость между продолжительностью контракта игрока и его стоимость? 

In [5]:
pd.set_option('display.max_columns', None)

In [92]:
df

ID                Name  Age  \
0      158023            L. Messi   31   
1       20801   Cristiano Ronaldo   33   
2      190871           Neymar Jr   26   
3      193080              De Gea   27   
4      192985        K. De Bruyne   27   
...       ...                 ...  ...   
18202  238813        J. Lundstram   19   
18203  243165  N. Christoffersson   19   
18204  241638           B. Worman   16   
18205  246268      D. Walker-Rice   17   
18206  246269           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate 

Переведем столбцы Contract Valid Until и Joined в числа. Для упрощения работы будем отбрасмывать месяцы и дни, и работать только с годами.

In [93]:
def tonumber(x):
  listx = list(x.split(' '))
  return int(listx[-1])

In [94]:
df_new = df[df['Joined'].notna()]

In [95]:
df_new = df_new[df_new['Contract Valid Until'].notna()]

In [96]:
df_new['Joined int'] = df_new['Joined'].apply(tonumber)

In [97]:
df_new['Contract Valid Until int'] = df_new['Contract Valid Until'].apply(tonumber)

In [98]:
df_new[['Joined int', 'Contract Valid Until']]

Joined int Contract Valid Until
0            2004                 2021
1            2018                 2022
2            2017                 2022
3            2011                 2020
4            2015                 2023
...           ...                  ...
18202        2017                 2019
18203        2018                 2020
18204        2017                 2021
18205        2018                 2019
18206        2018                 2019

[16654 rows x 2 columns]

In [99]:
df_new['Duration'] = df_new['Contract Valid Until int'] - df_new['Joined int']

In [100]:
df_new['MultV']=np.where(df_new.Value.str[-1]=='K',1000,np.where(df_new.Value.str[-1]=='M',1000000,1))

In [101]:
df_new['Value']=df_new.Value.str.strip('€')
df_new['Value']=df_new.Value.str.strip('K')
df_new['Value']=df_new.Value.str.strip('M')
df_new['Value']=pd.to_numeric(df_new['Value'])
df_new['Value in €']=df_new['Value']*df_new['MultV']

In [102]:
df_new_ = df_new.groupby('Duration').apply(lambda x: x['Value in €'].median()).reset_index(name='Median Value')

In [105]:
fig=px.scatter(df_new_, x='Duration',y='Median Value',title='Dependency Value on Duration contract')
fig.show(renderer='colab')

Да, зависимость есть - цена повышается к 0-15 годам контракта, а потом падает

# В какой год будет больше всего свободных игроков на рынке (исходя из этих данных)? визуализируйте зависимость 

In [109]:
fig = px.histogram(df_new, x='Contract Valid Until int', color='Contract Valid Until int')
fig.show(renderer='colab')

В 2019

# Есть ли зависмость потенциала (Potential) игрока от возраста? Всегда ли молодые игроки обладают большим потенциалом? 

In [ ]:
df_new = df.groupby('Age').apply(lambda x: x['Potential'].median()).reset_index(name='Median Potential')

In [ ]:
fig=px.scatter(df_new, x='Age',y='Median Potential',title='Dependency Median Potential on Age')
fig.show(renderer='colab')

Да, с возрастом медианное значение потенциала уменьшается, но не очень сильно

# Есть ли зависимость рейтинга игрока (Overall) от возраста? Когда у игроков "пик" карьеры? Игроков какого возраста больше всего с рейтингом 90+?  

In [16]:
df_new = df.groupby('Age').apply(lambda x: x['Overall'].median()).reset_index(name='Median Overall')

In [17]:
fig=px.scatter(df_new, x='Age',y='Median Overall',title='Dependency Median Overall on Age')
fig.show(renderer='colab')

Да, рейтинг увеличивается с возрастом. Пик карьеры - в районе 30 лет.

In [23]:
df_top = df[df['Overall'] >= 90]

In [25]:
df_top = df_top.groupby('Age').apply(lambda x: x['Name'].count()).reset_index(name='Counts')

In [26]:
df_top

Age  Counts
0   25       1
1   26       1
2   27       3
3   28       1
4   29       1
5   31       2
6   32       4
7   33       1

In [28]:
fig=px.bar(df_top, x='Age',y='Counts',title='Number of highly rated players by age')
fig.show(renderer='colab')

Игроков с рейтингом больше 90 больше всего среди 32-летних.

# Насколько средняя зарплата (Wage) игроков с рейтингом (Overall) 90+ больше средней зарплаты с рейтингом [80 - 90]? Визуализируйте возможную зависимость

In [60]:
wage_over=df[['Name','Wage','Overall']]
wage_over['MultW']=np.where(wage_over.Wage.str[-1]=='K',1000,1)
wage_over['Wage']=wage_over.Wage.str.strip('K')
wage_over['Wage']=wage_over.Wage.str.strip('€')
wage_over['Wage']=pd.to_numeric(wage_over['Wage'])
wage_over['Wage in €']=wage_over['Wage']*wage_over['MultW']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/st

In [34]:
wage_90 = wage_over[wage_over['Overall'] >= 90].Wage.mean()

In [35]:
wage_80_90 = wage_over[(wage_over['Overall'] <= 90) & (wage_over['Overall'] >= 80)].Wage.mean()

In [36]:
wage_90 - wage_80_90

240.27472527472526

На 240 тысяч евро

Построим зависимость средней зарплаты от рейтинга:

Разобьем рейтинг на отрезки по 10:

In [63]:
wage_over['Overall by 10 points'] = wage_over['Overall'] // 10

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [64]:
df_new = wage_over.groupby('Overall by 10 points').apply(lambda x: x['Wage'].mean()).reset_index(name='Mean wage')

In [66]:
fig=px.bar(df_new, x='Overall by 10 points',y='Mean wage',title='Wage dependence on the Overall rating')
fig.show(renderer='colab')